src/sync/PushManager/index.js

```diff
@@ -0,0 +1,201 @@
+const log = logFactory('splitio-sync:push-manager');
+import SplitUpdateWorker from '../SplitUpdateWorker';
+import SegmentUpdateWorker from '../SegmentUpdateWorker';
+import checkPushSupport from './checkPushSupport';
+import Backoff from '../../utils/backoff';
+import { hashUserKey } from '../../utils/jwt/hashUserKey';
+import EventEmitter from 'events';
+import { SECONDS_BEFORE_EXPIRATION, PUSH_DISCONNECT, PUSH_DISABLED, SSE_ERROR, SPLIT_KILL, SPLIT_UPDATE, SEGMENT_UPDATE, MY_SEGMENTS_UPDATE } from '../constants';
+
+/**
+ * Factory of the push mode manager.
+ *
+ * @param {Object} context context of main client.
+ * @param {Object | undefined} clientContexts map of user keys to client contexts to handle sync of MySegments. undefined for node.
+ */
+export default function PushManagerFactory(context, clientContexts /* undefined for node */) {
+
+  // No return a PushManager if PUSH mode is not supported.
+  if (!checkPushSupport(log))
+    return;
+
+  const pushEmitter = new EventEmitter();
+
+  const settings = context.get(context.constants.SETTINGS);
+  const storage = context.get(context.constants.STORAGE);
+  const sseClient = SSEClient.getInstance(settings);
+  const sseHandler = SSEHandlerFactory(pushEmitter);
+  sseClient.setEventHandler(sseHandler);
+
+  // map of hashes to user keys, to dispatch MY_SEGMENTS_UPDATE events to the corresponding MySegmentsUpdateWorker
+  const userKeyHashes = {};
+  // variable used on browser to reconnect only when a new client was added, saving some authentication and sse connections.
+  let connectForNewClient = false;
+
+  /** PushManager functions related to initialization */
+
+  const reauthBackoff = new Backoff(connectPush, settings.authRetryBackoffBase);
+  const sseReconnectBackoff = new Backoff(sseClient.reopen.bind(sseClient), settings.streamingReconnectBackoffBase);
+
+  let timeoutId = 0;
+
+  function scheduleTokenRefresh(issuedAt, expirationTime) {
+    // clear scheduled token refresh if exists (needed when resuming PUSH)
+    if (timeoutId) clearTimeout(timeoutId);
+
+    // Set token refresh 10 minutes before expirationTime
+    const delayInSeconds = expirationTime - issuedAt - SECONDS_BEFORE_EXPIRATION;
+
+    timeoutId = setTimeout(connectPush, delayInSeconds * 1000);
+  }
+
+  function connectPush() {
+    log.info('Connecting to push streaming.');
+
+    const userKeys = clientContexts ? Object.keys(clientContexts) : undefined;
+    authenticate(settings, userKeys).then(
+      function (authData) {
+        // restart backoff retry counter for auth and SSE connections, due to HTTP/network errors
+        reauthBackoff.reset();
+        sseReconnectBackoff.reset(); // reset backoff in case SSE conexion has opened after a HTTP or network error.
+
+        // emit PUSH_DISCONNECT if org is not whitelisted
+        if (!authData.pushEnabled) {
+          log.error('Streaming is not enabled for the organization. Switching to polling mode.');
```

Comment:
Use a comment in the code to mark it for review before releasing.

For now, use something like: "Streaming is not available. Switching to polling mode."

don't use error level, because it'll create lots of noise during the rollout phases. Go with INFO.

Reply:
Emi, let's please make sure we change this before merging.